### PR TITLE
M-011: implement TS auth retry policy

### PR DESCRIPTION
**Pull request by lsm**
## Summary
- Add typed `MakaiAuthRequiredError` and normalize manual `auth_required` failures with provider IDs.
- Update provider/agent auth retry handling so `auto_once` logs in once, retries once, preserves handler precedence, and fails fast when interaction needs missing handlers.
- Expand auth event flattening and retry-policy coverage in TS tests.

## Test plan
- [x] `npm test`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4 (Anthropic)

> **Model:** Claude Opus 4 | **Client:** NeoKai | **Provider:** Anthropic

### Summary

The implementation is well-structured overall: `MakaiAuthRequiredError` is properly typed, exported, and used across all four code paths (provider complete/stream, agent run/stream). Auth event flattening tests now cover all five wire variants. The `FakeTransport`/`FakeAuth` test pattern in `chat_client.test.ts` is clean and effectively isolates retry behavior. All 93 unit tests pass.

**Verdict: REQUEST_CHANGES** — one behavioral inconsistency and one test gap need to be addressed before merging.

---

### P2-001: `withAuthRetry` does not wrap post-retry `auth_required` as `MakaiAuthRequiredError`

**File:** `typescript/src/execution_client.ts`, line 928

In `withAuthRetry`, the retried operation at `return await operation()` is **not** wrapped in a try-catch. If the retried request *also* returns `auth_required`, the raw `MakaiStreamError` propagates out instead of being wrapped as `MakaiAuthRequiredError`.

This is inconsistent with:
- The **streaming** paths (provider `stream()` line 138–140, agent `stream()` line 279–281) which correctly convert all post-retry `auth_required` errors to `MakaiAuthRequiredError` via `authRequiredError()`.
- The **manual** policy path (line 930–932) which also wraps `auth_required` as `MakaiAuthRequiredError`.

**Expected:** After `auto_once` login succeeds but the retried request still fails with `auth_required`, the error should be `MakaiAuthRequiredError` — same as every other `auth_required` exit path.

**Suggested fix** — wrap the retried operation in a try-catch that normalizes `auth_required` errors:

```typescript
// In withAuthRetry, replace line 928:
options.beforeRetry?.();
try {
  return await operation();
} catch (retryError) {
  if (isRetryableAuthError(retryError)) {
    const providerId = retryError.provider_id ?? options.fallbackProviderId;
    if (providerId) throw authRequiredError(providerId, retryError.message);
  }
  throw retryError;
}
```

---

### P2-002: Missing test — non-streaming `auto_once` where retry still returns `auth_required`

**File:** `typescript/test/chat_client.test.ts` (or `execution_client.test.ts`)

The streaming paths have tests for "at most once" behavior:
- `client.provider.stream auto_once retries at most once when auth_required persists`
- `client.agent.stream auto_once retries at most once when auth_required persists`

But there is **no equivalent test** for the non-streaming paths (`provider.complete` / `agent.run`). Adding one in `chat_client.test.ts` would also cover P2-001 after the fix is applied (verify the error is `MakaiAuthRequiredError`, not raw `MakaiStreamError`).

---

### Minor observations (no action required)

- **`isInteractiveAuthWithoutHandlers` regex coupling** (line 880): The detection relies on regex-matching the auth client's error message string. If the message changes, the detection breaks silently. A more robust approach would be to add a sentinel `code` or property to `MakaiAuthError`. This is a low-priority improvement and not blocking.

- **Existing integration tests** in `execution_client.test.ts` check for `MakaiStreamError` rather than `MakaiAuthRequiredError`. Since `MakaiAuthRequiredError extends MakaiStreamError`, these tests pass, but they don't verify the specific subtype. Not blocking — the new unit tests in `chat_client.test.ts` already cover this correctly.

File: typescript/src/execution_client.ts [hunk 1]
```diff
@@ -895,12 +918,19 @@ async function withAuthRetry<T>(
...
-      } catch {
-        throw error;
+      } catch (loginError) {
+        if (options.authHandlers === undefined && isInteractiveAuthWithoutHandlers(loginError)) {
+          throw authRequiredError(providerId, error.message);
+        }
+        throw loginError;
       }
       options.beforeRetry?.();
       return await operation();
```
Comment (lsm) [inline typescript/src/execution_client.ts]: **P2-001:** The retried `return await operation()` here is not wrapped in a try-catch. If the retried operation also returns `auth_required`, the raw `MakaiStreamError` propagates instead of `MakaiAuthRequiredError`. This is inconsistent with the streaming paths (lines 138–140, 279–281) and the manual policy path (lines 930–932). See the review body for a suggested fix.

Comment (lsm) [reply typescript/src/execution_client.ts]: Fixed in 7b42a35. The auto_once retry attempt is now wrapped and any repeated auth_required is normalized to MakaiAuthRequiredError. I also added non-streaming retry-limit tests for provider.complete and agent.run to cover this case.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4 (Anthropic)

> **Model:** Claude Opus 4 | **Client:** NeoKai | **Provider:** Anthropic

### Round 2 — APPROVE

Both P2 findings from round 1 have been addressed:

1. **P2-001 fixed** — `withAuthRetry` (line 928–936) now wraps the retried `operation()` in a try-catch. If the retry returns `auth_required`, it's correctly converted to `MakaiAuthRequiredError` with proper provider_id fallback. This makes all four code paths (provider complete/stream, agent run/stream) consistent.

2. **P2-002 fixed** — Two new integration tests added:
   - `client.provider.complete auto_once retries at most once when auth_required persists`
   - `client.agent.run auto_once retries at most once when auth_required persists`
   
   Both verify the error is specifically `MakaiAuthRequiredError` (not just `MakaiStreamError`), covering the previously untested non-streaming retry-failure case.

All 95 tests pass (5 e2e skipped as expected). Implementation is clean, consistent, and complete.

**Comment by lsm** [comment]
@codex review

File: typescript/src/execution_client.ts [hunk 2]
```diff
@@ -895,11 +918,26 @@ async function withAuthRetry<T>(
       if (!providerId) throw error;
       try {
         await options.auth.login(providerId, options.authHandlers);
-      } catch {
-        throw error;
+      } catch (loginError) {
+        if (options.authHandlers === undefined && isInteractiveAuthWithoutHandlers(loginError)) {
+          throw authRequiredError(providerId, error.message);
```
[the remark this answers is not in the document]
Comment (lsm) [reply typescript/src/execution_client.ts]: Fixed in dcffb14. auto_once now preserves the original auth_required contract for any login failure, so callers receive MakaiAuthRequiredError with provider_id regardless of whether handlers are absent, partial (e.g. onEvent only), or login fails for another reason. This also removes the brittle message-regex classification entirely.

[the remark this answers is not in the document]
Comment (lsm) [reply typescript/src/execution_client.ts]: Fixed in dcffb14. auto_once now preserves the original auth_required contract for any login failure, so callers receive MakaiAuthRequiredError with provider_id regardless of whether handlers are absent, partial (e.g. onEvent only), or login fails for another reason. This also removes the brittle message-regex classification entirely.

File: typescript/src/execution_client.ts [hunk 3]
```diff
@@ -857,6 +872,14 @@ function isRetryableAuthError(error: unknown): error is MakaiStreamError {
   return error instanceof MakaiStreamError && error.code === "auth_required";
 }
 
+function authRequiredError(providerId: string, message: string): MakaiAuthRequiredError {
+  return new MakaiAuthRequiredError(providerId, message);
+}
+
+function isInteractiveAuthWithoutHandlers(error: unknown): boolean {
+  return error instanceof Error && /no onPrompt handler configured/.test(error.message);
```
[the remark this answers is not in the document]
Comment (lsm) [reply typescript/src/execution_client.ts]: Fixed in dcffb14. auto_once now preserves the original auth_required contract for any login failure, so callers receive MakaiAuthRequiredError with provider_id regardless of whether handlers are absent, partial (e.g. onEvent only), or login fails for another reason. This also removes the brittle message-regex classification entirely.

File: typescript/src/execution_client.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (lsm) [reply typescript/src/execution_client.ts]: Fixed in dcffb14. auto_once now preserves the original auth_required contract for any login failure, so callers receive MakaiAuthRequiredError with provider_id regardless of whether handlers are absent, partial (e.g. onEvent only), or login fails for another reason. This also removes the brittle message-regex classification entirely.

File: typescript/src/execution_client.ts [hunk 4]
```diff
@@ -122,15 +123,22 @@ class StdioProviderApi implements MakaiProviderApi {
           if (providerId) {
             try {
               await this.auth.login(providerId, this.authHandlers);
-            } catch {
-              throw error;
+            } catch (loginError) {
+              if (this.authHandlers === undefined && isInteractiveAuthWithoutHandlers(loginError)) {
+                throw authRequiredError(providerId, error.message);
```
[the remark this answers is not in the document]
Comment (lsm) [reply typescript/src/execution_client.ts]: Fixed in dcffb14. auto_once now preserves the original auth_required contract for any login failure, so callers receive MakaiAuthRequiredError with provider_id regardless of whether handlers are absent, partial (e.g. onEvent only), or login fails for another reason. This also removes the brittle message-regex classification entirely.

File: typescript/src/execution_client.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (lsm) [reply typescript/src/execution_client.ts]: Fixed in dcffb14. auto_once now preserves the original auth_required contract for any login failure, so callers receive MakaiAuthRequiredError with provider_id regardless of whether handlers are absent, partial (e.g. onEvent only), or login fails for another reason. This also removes the brittle message-regex classification entirely.

File: typescript/src/execution_client.ts [hunk 5]
```diff
@@ -895,11 +918,26 @@ async function withAuthRetry<T>(
       if (!providerId) throw error;
       try {
         await options.auth.login(providerId, options.authHandlers);
-      } catch {
-        throw error;
+      } catch (loginError) {
+        if (options.authHandlers === undefined && isInteractiveAuthWithoutHandlers(loginError)) {
+          throw authRequiredError(providerId, error.message);
+        }
+        throw loginError;
```
[the remark this answers is not in the document]
Comment (lsm) [reply typescript/src/execution_client.ts]: Fixed in dcffb14. auto_once now preserves the original auth_required contract for any login failure, so callers receive MakaiAuthRequiredError with provider_id regardless of whether handlers are absent, partial (e.g. onEvent only), or login fails for another reason. This also removes the brittle message-regex classification entirely.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4 (Anthropic)

> **Model:** Claude Opus 4 | **Client:** NeoKai | **Provider:** Anthropic

### Round 3 — APPROVE

Previous P2 findings fully addressed with an even cleaner approach than the previous round:

1. **P2-001 (retry error normalization):** `withAuthRetry` now wraps the retried `operation()` in try-catch (lines 915–923), converting `auth_required` errors to `MakaiAuthRequiredError` with proper provider_id fallback.

2. **P2-002 (test coverage for non-streaming retry-failure):** Two new integration tests added — `client.provider.complete auto_once retries at most once when auth_required persists` and `client.agent.run auto_once retries at most once when auth_required persists` — both verifying `MakaiAuthRequiredError` specifically.

3. **Simplification:** The fragile `isInteractiveAuthWithoutHandlers` regex helper was removed entirely. All login failures during `auto_once` now consistently normalize to `MakaiAuthRequiredError` regardless of cause — simpler, more robust, and consistent across all four code paths.

4. The renamed test `auto_once normalizes login failure to auth_required with partial handlers` properly validates the partial-handler edge case (has `onEvent` but not `onPrompt`).

All 8 review threads resolved. All 95 tests pass. Clean approval.
